rooms.js

```diff
@@ -163,6 +166,142 @@ var Room = (function () {
 		}
 		return true;
 	};
+	//mute handling
+	Room.prototype.runMuteTimer = function () {
+		if (this.mutedUsers[0].time < Date.now() + 5) {
+			// this prevents lag from bugging the timer
+			this.mutedUsers[0].time = Date.now() + 5;
+		}
+		var self = this;
+		this.muteTimer = setTimeout(function () {
+			var user = Users(self.mutedUsers[0].userid);
+			if (user) {
+				user.popup("Your mute has expired.");
+				self.unMute(user);
+			}
+		}, self.mutedUsers[0].time - Date.now());
+	};
+	Room.prototype.isMuted = function (user) {
+		if (!user) return;
+		if (this.mutedUsers) {
+			for (var i = 0; i < this.mutedUsers.length; i++) {
+				if (user.userid === this.mutedUsers[i].userid) {
+					return this.mutedUsers[i].userid;
+				}
+				if (user.autoconfirmed === this.mutedUsers[i].userid) {
+					return this.mutedUsers[i].userid;
+				}
+			}
+		}
+		if (this.mutedIps) {
+			for (var ip in user.ips) {
+				if (ip in this.mutedIps) return this.mutedIps[ip];
+			}
+		}
+	};
+	Room.prototype.getMuteTime = function (user) {
+		var userid = this.isMuted(user);
+		if (!userid) return;
+		for (var i = 0; i < this.mutedUsers.length; i++) {
+			if (userid === this.mutedUsers[i].userid) {
+				return this.mutedUsers[i].time - Date.now();
+			}
+		}
+	};
+	Room.prototype.mute = function (user, setTime, noRecurse) {
+		var userid = user.userid;
+		var alts;
+		if (!noRecurse) {
+			alts = [];
+			for (var i in Users.users) {
+				var otherUser = Users.users[i];
+				if (otherUser === user) continue;
+				for (var myIp in user.ips) {
+					if (myIp in otherUser.ips) {
+						alts.push(otherUser.name);
+						this.mute(otherUser, setTime, true);
+						break;
+					}
+				}
+			}
+		}
+
+		if (!setTime) setTime = 7 * 60000; // default time: 7 minutes
+		if (setTime < 1) setTime = 1; // mostly to prevent bugs
+		if (setTime > 90 * 60000) setTime = 90 * 60000; // limit 90 minutes
+
+		// Place the user in a queue for the unmute timer
+		for (var queue = 0; queue < this.mutedUsers.length + 1; queue++) {
+			var timeRemaining = this.mutedUsers[queue] ? this.mutedUsers[queue].time - Date.now() : Infinity;
+			if (setTime < timeRemaining) {
+				var buffer = Object.create(null);
+				buffer.userid = userid;
+				buffer.time = Date.now() + setTime;
+				this.mutedUsers.splice(queue, 0, buffer);
+				break;
+			}
+		}
+		if (!this.muteTimer) {
+			this.runMuteTimer();
```

Comment:
Instead of `if (!this.muteTimer) {`, you should just call `this.runMuteTimer();` every time, and then have

    if (this.muteTimer) return;

at the top of `runMuteTimer`.